odoo/addons/base/models/ir_actions.py

```diff
@@ -117,25 +152,22 @@ def _get_bindings(self, model_name, debug=False):
         for action_id, action_model, binding_type in cr.fetchall():
             try:
                 action = self.env[action_model].sudo().browse(action_id)
-                action_groups = getattr(action, 'groups_id', ())
-                action_model = getattr(action, 'res_model', False)
-                if action_groups and not action_groups & user_groups:
-                    # the user may not perform this action
-                    continue
-                if action_model and not IrModelAccess.check(action_model, mode='read', raise_exception=False):
-                    # the user won't be able to read records
-                    continue
                 fields = ['name', 'binding_view_types']
-                if 'sequence' in action._fields:
-                    fields.append('sequence')
-                result[binding_type].append(action.read(fields)[0])
-            except (AccessError, MissingError):
+                for field in ('groups_id', 'res_model', 'sequence'):
+                    if field in action._fields:
+                        fields.append(field)
+                action = action.read(fields)[0]
+                if action.get('groups_id'):
+                    groups = self.env['res.groups'].browse(action['groups_id'])
+                    action['groups_id'] = ','.join(ext_id for ext_id in groups.get_external_id().values() if ext_id)
```

Comment:
`_ensure_group_xml_id` seems to be called in `create` and `write`.

Is there a migration for existing groups? Also is the `write` call really necessary?

Reply:
for `write` it is necessary, in case a user manually adds or changes a group on an action. The group added or changed to must have an external identifier.

Though I was also considering changing the signature of `_ensure_group_xml_id` to return a dict with group id as key and external identifier name as value, as `_get_external_id` is doing, but creating the external identifier on the fly if not existing, and call `_ensure_group_xml_id` instead of `_get_external_id` in `_get_bindings` directly.

That way, no need to add `_ensure_group_xml_id` in create nor write, and no need of upgrade script. But it means a method which is supposed to only "read" things will potentially start writing things, which is a bit weird but conceivable.
What do you think ?

For the record, there is a similar method existing in `models.py`
https://github.com/odoo/odoo/blob/216b1a1b2f79016922c433692f82974fa00f6133/odoo/models.py#L838
that I cannot use because of the double `__` in the method prefix.

Reply:
> For the record, there is a similar method existing in models.py

Kinda, that one's got a pretty specific role as it's a bulk xid-er for the export, hence doing all sorts of weird stuff in order to use `COPY`.